subgraphs/_reference_/src/sdk/lending/snapshots.ts

```diff
@@ -0,0 +1,717 @@
+  }
+
+  createOrUpdateUsageDailySnapshot(): void {
+    const days = this.event.block.timestamp.toI32() / SECONDS_PER_DAY;
+    const id = Bytes.fromI32(days);
+    let snapshot = UsageMetricsDailySnapshot.load(id);
+
+    if (!snapshot) {
+      snapshot = new UsageMetricsDailySnapshot(id);
+      snapshot.days = days;
+      snapshot.protocol = this.protocol.id;
+      snapshot.dailyActiveUsers = INT_ZERO;
+      snapshot.dailyActiveDepositors = INT_ZERO;
+      snapshot.dailyActiveBorrowers = INT_ZERO;
+      snapshot.dailyActiveLiquidators = INT_ZERO;
+      snapshot.dailyActiveLiquidatees = INT_ZERO;
+      snapshot.dailyTransactionCount = INT_ZERO;
+      snapshot.dailyDepositCount = INT_ZERO;
+      snapshot.dailyWithdrawCount = INT_ZERO;
+      snapshot.dailyBorrowCount = INT_ZERO;
+      snapshot.dailyRepayCount = INT_ZERO;
+      snapshot.dailyLiquidateCount = INT_ZERO;
+      snapshot.dailyTransferCount = INT_ZERO;
+      snapshot.dailyFlashloanCount = INT_ZERO;
+      snapshot.dailyActivePositions = INT_ZERO;
+    }
+    snapshot.cumulativeUniqueUsers = this.protocol.cumulativeUniqueUsers;
+    snapshot.cumulativeUniqueDepositors =
+      this.protocol.cumulativeUniqueDepositors;
+    snapshot.cumulativeUniqueBorrowers =
+      this.protocol.cumulativeUniqueBorrowers;
+    snapshot.cumulativeUniqueLiquidators =
+      this.protocol.cumulativeUniqueLiquidators;
+    snapshot.cumulativeUniqueLiquidatees =
+      this.protocol.cumulativeUniqueLiquidatees;
+    snapshot.cumulativePositionCount = this.protocol.cumulativePositionCount;
+    snapshot.openPositionCount = this.protocol.openPositionCount;
+    snapshot.totalPoolCount = this.protocol.totalPoolCount;
+    snapshot.blockNumber = this.event.block.number;
+    snapshot.timestamp = this.event.block.timestamp;
+
+    snapshot.save();
+    this.usageDailySnapshot = snapshot;
+  }
+
+  createOrUpdateUsageHourlySnapshot(): void {
+    const hours = this.event.block.timestamp.toI32() / SECONDS_PER_HOUR;
+    const id = Bytes.fromI32(hours);
+    let snapshot = UsageMetricsHourlySnapshot.load(id);
+
+    if (!snapshot) {
+      snapshot = new UsageMetricsHourlySnapshot(id);
+      snapshot.hours = hours;
+      snapshot.protocol = this.protocol.id;
+      snapshot.hourlyActiveUsers = INT_ZERO;
+      snapshot.hourlyTransactionCount = INT_ZERO;
+      snapshot.hourlyDepositCount = INT_ZERO;
+      snapshot.hourlyWithdrawCount = INT_ZERO;
+      snapshot.hourlyBorrowCount = INT_ZERO;
+      snapshot.hourlyRepayCount = INT_ZERO;
+      snapshot.hourlyLiquidateCount = INT_ZERO;
+    }
+    snapshot.cumulativeUniqueUsers = this.protocol.cumulativeUniqueUsers;
+    snapshot.blockNumber = this.event.block.number;
+    snapshot.timestamp = this.event.block.timestamp;
+
+    snapshot.save();
+    this.usageHourlySnapshot = snapshot;
+  }
+
+  ////////////////////
+  ///// Updaters /////
+  ////////////////////
+
+  public updateUsageData(transactionType: string, account: Bytes): void {
```

Comment:
I suggest breaking this function down. It is easy to forget of some step when making an update, or assigning values to the wrong counters. In the end it is the same action being repeated for each activity period.

Something like this would ensure that all actions (create ID, attempt to load, save if not exist updating counters) are always performed together:
```
registerActivity(period: AccountActivity, transactionType: string, account: Bytes, market: Bytes, timestamp: BigInt): void {
  // generate ID
  // attempt to load
  // update counter and save 
}
```

Or if you don’t want to have a giant `if` inside to choose which counter to update, have it return either 1 or 0 and add that value to the counter you want:

```
this.usageDailySnapshot.dailyActiveDepositors += registerActivity(DAILY, DEPOSIT, account, “”, days);
this.marketDailySnapshot.dailyActiveUsers += registerActivity(DAILY, “”, account, this.market.id, days);
// ...
```

Reply:
fixed